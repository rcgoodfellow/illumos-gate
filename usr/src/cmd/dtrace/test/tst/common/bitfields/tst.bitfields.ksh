#!/usr/bin/ksh
#
#
# This file and its contents are supplied under the terms of the
# Common Development and Distribution License ("CDDL"), version 1.0.
# You may only use this file in accordance with the terms of version
# 1.0 of the CDDL.
#
# A full copy of the text of the CDDL should have accompanied this
# source.  A copy of the CDDL is also available via the Internet at
# http://www.illumos.org/license/CDDL.
#

#
# Copyright 2022 Oxide Computer Company
#

#
# This test acts as a series of regression tests in DTrace around
# printing bitfields that are byte sized at non-byte aligned offsets and
# around printing fields that are less than a byte in size, but due to their
# offset, cross a byte boundary (e.g. a 5-bit bitfield that starts at
# bit 6).
#
# The DTrace implementation has two different general paths for this:
#
#  o The D compiler compiling a dereference into DIF code to be executed
#    in probe context to extract a bitfield value.
#  o The print() action which grabs the larger chunk of memory and then
#    processes it all in userland.
#

if [ $# != 1 ]; then
        echo expected one argument: '<'dtrace-path'>'
        exit 2
fi

dtrace=$1
exe="tst.bitfields.exe"

elfdump "./$exe" | grep -q '.SUNW_ctf'
if (( $? != 0 )); then
	echo "CTF does not exist in $exe, that's a bug" >&2
	exit 1
fi

$dtrace -qs /dev/stdin -c "./$exe 0xe417 0x9391d7db" <<EOF
pid\$target::mumble:entry
{
	print(*args[1]);
	printf("\n");
	print(*args[2]);
	printf("\n");
	trace(args[2]->b);
	printf("\n0x%x 0x%x 0x%x 0x%x 0x%x 0x%x\n", args[2]->a, args[2]->b,
	    args[2]->c, args[2]->d, args[2]->e, args[2]->f);
	trace(args[1]->i);
	printf("\n0x%x 0x%x 0x%x 0x%x\n", args[1]->g, args[1]->h, args[1]->i,
	    args[1]->j);
}
EOF

exit $rc
