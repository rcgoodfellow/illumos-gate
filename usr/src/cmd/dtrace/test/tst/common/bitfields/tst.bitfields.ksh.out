bit0_t {
    unsigned int:3 a :3 = 0x3
    unsigned int:2 b :2 = 0x3
    unsigned int:1 c :1 = 0
    unsigned int:1 d :1 = 0x1
    unsigned int:1 e :1 = 0x1
    unsigned int:1 f :1 = 0x1
    unsigned int:3 g :3 = 0x3
    unsigned int:3 h :3 = 0x5
    unsigned int:5 i :5 = 0x3
    unsigned int:4 j :4 = 0x9
    unsigned int:6 k :6 = 0x13
    unsigned int:1 l :1 = 0
    unsigned int:1 m :1 = 0x1
}
bit1_t {
    unsigned short:1 a :1 = 0x1
    unsigned short:8 b :8 = 0xb
    unsigned short:3 c :3 = 0x2
    unsigned short:2 d :2 = 0x2
    unsigned short:1 e :1 = 0x1
    unsigned short:1 f :1 = 0x1
}
11
0x1 0xb 0x2 0x2 0x1 0x1
3
0x3 0x5 0x3 0x9

